### PR TITLE
MTS-3250/BR-3699: Upgrade task is interrupted on qa2

### PR DESCRIPTION
**Pull request by yellowandy**

### CODE REVIEW

File: shadow.c [hunk 1]
```diff
@@ -607,6 +607,9 @@ static char *template_to_instance(char *filename, int options TSRMLS_DC)
 		if(SHADOW_G(debug) & SHADOW_DEBUG_PATHCHECK) fprintf(stderr, "In template: %s\n", filename);
 		if((options & OPT_CHECK_EXISTS) && shadow_cache_get(filename, &newname) == SUCCESS) {
 			if(SHADOW_G(debug) & SHADOW_DEBUG_PATHCHECK) fprintf(stderr, "Path check from cache: %s => %s\n", filename, newname);
+			if(realpath) {
+            			efree(realpath);
```
Comment (skymeyer) [inline shadow.c:611]: Interesting how almost no (e)malloc happens in this whole code.

**Comment by morozov** [comment]
It seems to solve the problem. I checked the test script attached to the issue, and the memory usage no more depends on the number of iterations (20K, 200K, 500K).

I also profiled post upgrade scripts individually and the whole post stage. The memory used by `7_MergeDropdowns` went down from 300M to 10M, the whole post upgrade stage takes ~220M.

File: shadow.c (hunk 1 above)
Comment (dwheeler) [inline shadow.c:611]: It does implicitly through `estrndup`...

Makes me wonder if there are other leaks from this as well.

**Comment by dwheeler** [comment]
:+1:

**Pull request merged**